core/kernel/ldelf_syscalls.c

```diff
@@ -50,11 +52,26 @@ TEE_Result ldelf_syscall_map_zi(vaddr_t *va, size_t num_bytes, size_t pad_begin,
 	fobj_put(f);
 	if (!mobj)
 		return TEE_ERROR_OUT_OF_MEMORY;
-	res = vm_map_pad(uctx, va, num_bytes, prot, vm_flags,
+
+	res = GET_USER(_va, va);
+	if (res)
+		return res;
+
+	res = vm_map_pad(uctx, &_va, num_bytes, prot, vm_flags,
 			 mobj, 0, pad_begin, pad_end, 0);
 	mobj_put(mobj);
 
-	return res;
+	res = PUT_USER(_va, va);
+	if (res) {
+		if (vm_unmap(uctx, _va, num_bytes))
+			panic();
+
+		vm_set_ctx(uctx->ts_ctx);
```

Comment:
is it needed to call `vm_set_ctx(uctx->ts_ctx)`? i don't think so. Unmap instruction from line 66 is already effective once the function return with success.

Reply:
or maybe i'm wrong. @jenswi-linaro?